### PR TITLE
Revisit functions startups again

### LINKED ISSUE #158: Revisit applying startups via type attributes
While looking into #154, we investigated the possibility of having a single FunctionsStartupAttribute that could be applied to types as well as assemblies, so you could decorate the function.

That leads to a chicken and egg situation - startups can change the configuration, and the configuration can specify the function, which would in turn change which startups were applied.

But the frustrating thing is that in _almost all cases_ the function we expect to run before starting any complex processing is going to be the one we _actually_ run. So maybe we should go for a "trust and validate" model? Guess what the function target type is, include the startups from that function, and then if later on we determine that was incorrect (and that it would have changed which startups were invoked) we can throw an exception with a detailed explanation.

cc @amanda-tarafa - we'll want to discuss this.

### PR DESCRIPTION
**Pull request by jskeet**
This removes the test-specific attribute, allowing the same attribute to be used from both tests and production code, and for it to be specified at an assembly and type level.

There's a certain amount of fiddliness here which is only trying to address a very cornery corner case - but it's all in implementation rather than API.

None of this change is breaking compared with beta01, although it's breaking compared with "head" (so I'm glad we haven't released a beta02 yet).

Fixes #158

### CODE REVIEW

**Comment by jskeet** [comment]
Oh, there's a very small breaking change in terms of UseFunctionsStartups now taking a type as well as an assembly, but that's all :)

File: src/Google.Cloud.Functions.Testing/FunctionTestServer.cs [hunk 1]
```diff
@@ -47,12 +48,12 @@ public FunctionTestServer(Type functionTarget) : this(functionTarget, null)
...
         /// <summary>
         /// Creates a FunctionTestServer for the specified function type, with an optional
-        /// type to inspect for <see cref="FunctionTestStartupAttribute"/> attributes.
+        /// type to inspect for <see cref="FunctionsStartupAttribute"/> attributes.
         /// </summary>
         /// <param name="functionTarget">The function type to host in the server.</param>
-        /// <param name="attributedStartupType">A type to examine for <see cref="FunctionTestStartupAttribute"/> attributes to provide further customization.
-        /// If this is null, or no attributes are specified in the type (or its base class hierarchy) the startup classes in the
-        /// function assembly are used.</param>
+        /// <param name="attributedStartupType">A type to examine for <see cref="FunctionsStartupAttribute"/> attributes to provide further customization.
```
Comment (amanda-tarafa) [inline src/Google.Cloud.Functions.Testing/FunctionTestServer.cs:54]: nit: The name of this param seems backwards to me. Maybe startupSourceType? typeWithStartups? maybe even startupAttributedType? As it is it just looks like we are passing the Startup type.

File: src/Google.Cloud.Functions.Testing/FunctionTestServer.cs [hunk 2]
```diff
@@ -47,12 +48,12 @@ public FunctionTestServer(Type functionTarget) : this(functionTarget, null)
...
+        /// If this is null, or no attributes are specified in the type (or its base class hierarchy) or the assembly it's declared in,
+        /// the startup classes in the function assembly are used.</param>
```
Comment (amanda-tarafa) [inline src/Google.Cloud.Functions.Testing/FunctionTestServer.cs:56]: Not just the ones in the function assembly right? Also the ones specified via attributes in the function itself? Maybe, "the function's usual startup classes will be used".

File: src/Google.Cloud.Functions.Testing/FunctionTestServer.cs [hunk 3]
```diff
@@ -127,11 +128,11 @@ public FunctionTestServer() : this(null)
 
         /// <summary>
         /// Constructs a new instance serving <typeparamref name="TFunction"/>, with an optional
-        /// type to inspect for <see cref="FunctionTestStartupAttribute"/> attributes.
+        /// type to inspect for <see cref="FunctionsStartupAttribute"/> attributes.
         /// </summary>
-        /// <param name="attributedStartupType">A type to examine for <see cref="FunctionTestStartupAttribute"/> attributes to provide further customization.
-        /// If this is null, or no attributes are specified in the type (or its base class hierarchy) the startup classes in the
-        /// function assembly are used.</param>
+        /// <param name="attributedStartupType">A type to examine for <see cref="FunctionsStartupAttribute"/> attributes to provide further customization.
```
Comment (amanda-tarafa) [inline src/Google.Cloud.Functions.Testing/FunctionTestServer.cs:133]: nit: same here with the name of the parameter, if you decide to change it.

File: src/Google.Cloud.Functions.Testing/FunctionTestServer.cs [hunk 4]
```diff
@@ -127,11 +128,11 @@ public FunctionTestServer() : this(null)
...
+        /// If this is null, or no attributes are specified in the type (or its base class hierarchy) or the assembly it's declared in,
+        /// the startup classes in the function assembly are used.</param>
```
Comment (amanda-tarafa) [inline src/Google.Cloud.Functions.Testing/FunctionTestServer.cs:135]: Same comment as before re "the usual startups..."

File: src/Google.Cloud.Functions.Testing/FunctionTestServerBuilder.cs [hunk 5]
```diff
@@ -72,8 +72,9 @@ public FunctionTestServerBuilder UseFunctionsStartups(IEnumerable<FunctionsStart
         }
 
         /// <summary>
-        /// Examines <paramref name="attributedType"/> (and its base class hierarchy) for <see cref="FunctionTestStartupAttribute" />
-        /// annotations. If any such attributes are found, the startup classes specified by the attributes are used
+        /// Examines <paramref name="attributedType"/> (and its base class hierarchy, and the assembly in which it's declared)
+        /// for <see cref="FunctionsStartupAttribute" /> annotations.
+        /// If any such attributes are found, the startup classes specified by the attributes are used
         /// instead of the startup classes specified in the assembly containing the function target. If no attributes
```
Comment (amanda-tarafa) [inline src/Google.Cloud.Functions.Testing/FunctionTestServerBuilder.cs:78]: "instead of the usual startup classes for the target function..."

File: examples/Google.Cloud.Functions.Examples.IntegrationTests/TestableDependenciesTest.cs [hunk 6]
```diff
@@ -23,12 +23,12 @@ namespace Google.Cloud.Functions.Examples.IntegrationTests
 {
     /// <summary>
     /// By default, the test server will use the same Functions Startup classes
-    /// as normal. The TestFunctionStartup attribute specifies Functions Startup classes that
-    /// should be used for this test class, so that it can inject test dependencies instead of the production ones.
+    /// as normal. Applying the FunctionsStartup attribute to the test class (or the assembly) signals to FunctionTestBase
```
Comment (amanda-tarafa) [inline examples/Google.Cloud.Functions.Examples.IntegrationTests/TestableDependenciesTest.cs:26]: nit: (and/or the assembly)

File: src/Google.Cloud.Functions.Hosting/HostingInternals.cs [hunk 7]
```diff
@@ -84,6 +88,25 @@ internal static IApplicationBuilder ConfigureApplication(WebHostBuilderContext c
                 });
         }
 
+        /// <summary>
+        /// Validates that the startup classes which would be used for the finally-selected function are the
+        /// same as the startup classes we *actually* used. This method is only called when the app is started
```
Comment (amanda-tarafa) [inline src/Google.Cloud.Functions.Hosting/HostingInternals.cs:93]: ...are the same and in the same order...

File: docs/customization.md [hunk 8]
```diff
@@ -169,3 +199,40 @@ you genuinely intend to remove the Functions Framework console logging.
 
 Logging for tests is provided automatically in the `FunctionTestServer`; see [the testing
 documentation](testing.md) for more details.
+
+## Startup class mismatches
+
+If you haven't received an error referring you to this section of
+the documentation, you *may* want to skip this section.
+
+It is possible (but very rare) for startup classes to affect which
```
Comment (amanda-tarafa) [inline docs/customization.md:208]: Ahhh, there's a use case that could be until now, and now can't (a little far fetch though)

What if StartupA, say, depending on configuration, were to add another StartupB on it's own ConfigureServices? StartupB just overwrites the Configure method to add a few things to the pipeline later. That's sort of valid (and StartupB would have actually been executed properly) and sort of makes sense, but would fail now where it didn't before. Maybe we can allow for startup additions with the warning that, the ConfigureConfiguration override won't execute because it's already happened?

I'm actually torn between what should be invalid, a change in startups or a change in function type? Maybe later this can be configurable, like allowFunctionSwap or allowStartupAddition?

File: examples/Google.Cloud.Functions.Examples.AdvancedDependencyInjection/Function.cs [hunk 9]
```diff
@@ -72,14 +65,25 @@ public class Startup : FunctionsStartup
...
     /// The actual Cloud Function using the dependencies.
     /// This uses both IOperationSingleton and IOperationScoped, and outputs the OperationId
     /// from each of them in HandleAsync. The service configuration provided in Startup
     /// means that for any given server, the SingletonOperationId property remains the same, but
     /// ScopedOperationId varies for each request.
     /// (Restarting the server will result in a different SingletonOperationId - nothing is persisted
     /// here.)
+    /// </para>
+    /// <para>
+    /// The FunctionsStartup attribute specifies the startup class to load. If you have multiple
```
Comment (amanda-tarafa) [inline examples/Google.Cloud.Functions.Examples.AdvancedDependencyInjection/Function.cs:78]: nit: `<see ref=FunctionsStartupAttribute />`?

**Review by amanda-tarafa — APPROVED** [review]
LGTM. A few nits, and a couple of comments/questions but I'm fine with this as is.

File: src/Google.Cloud.Functions.Hosting/FunctionsStartupAttribute.cs [hunk 10]
```diff
@@ -43,5 +48,25 @@ public class FunctionsStartupAttribute : Attribute
...
+        /// <summary>
+        /// Returns the functions startup classes found in the attributes of <paramref name="assembly"/> and
+        /// <paramref name="target"/>, ordered by <see cref="Order"/> and then by the full name of the type.
+        /// </summary>
+        /// <param name="assembly">The assembly to inspect for attributes.</param>
+        /// <param name="target">Optional type to query for attributes. If this is null, only
+        /// startup classes from <paramref name="assembly"/> are returned.
+        /// </param>
+        /// <returns>A (possibly empty) sequence of the startup types detected via attributes.</returns>
+        public static IEnumerable<Type> GetStartupTypes(Assembly assembly, Type? target)
```
[the remark this answers is not in the document]
Comment (jskeet) [reply src/Google.Cloud.Functions.Hosting/FunctionsStartupAttribute.cs:61]: Yes, the assembly of the `target` should be the same as `assembly`. (If `target` were required, I'd drop `assembly` entirely.) At least, I can't think of any times when they'd be different for production code. For test code it's just about plausible, although we wouldn't use that ourselves.

Do you think it's worth validating that? We could always relax it later.

File: src/Google.Cloud.Functions.Hosting/HostingInternals.cs [hunk 11]
```diff
@@ -156,14 +157,14 @@ internal static IServiceCollection AddServicesForFunctionTarget(IServiceCollecti
         /// discover problems there easily enough anyway.)
         /// </summary>
         /// <remarks>
-        /// This method is internal and in this class rather than being private and in Builder for the sake of testability.
+        /// This method is internal rather private for the sake of testability.
```
[the remark this answers is not in the document]
Comment (jskeet) [reply src/Google.Cloud.Functions.Hosting/HostingInternals.cs:183]: Done.

File: src/Google.Cloud.Functions.Hosting/HostingInternals.cs [hunk 12]
```diff
@@ -172,7 +173,33 @@ internal static Type FindDefaultFunctionType(params Type[] types)
...
 
-            bool IsFunctionClass(Type t) =>
+        }
+
+        /// <summary>
+        /// Attempts to find a single valid non-abstract function class within the given set of types, but returns
+        /// null (instead of failing) if there isn't exactly one such class.
+        /// </summary>
+        /// <param name="types"></param>
+        /// <returns></returns>
```
[the remark this answers is not in the document]
Comment (jskeet) [reply src/Google.Cloud.Functions.Hosting/HostingInternals.cs:206]: Done.

File: src/Google.Cloud.Functions.Hosting/HostingInternals.cs [hunk 13]
```diff
@@ -113,15 +105,24 @@ internal static async Task Execute(HttpContext context)
...
                 : assembly.GetType(target) ?? throw new InvalidOperationException($"Can't load specified function type '{target}'");
         }
 
+        internal static Type? TryGetFunctionTarget(IConfiguration configuration, Assembly assembly)
+        {
+            var options = FunctionsFrameworkOptions.FromConfiguration(configuration);
+            string? target = options.FunctionTarget;
+            return target is null
+                ? TryFindDefaultFunctionType(assembly.GetTypes())
+                : assembly.GetType(target) ?? throw new InvalidOperationException($"Can't load specified function type '{target}'");
```
[the remark this answers is not in the document]
Comment (jskeet) [reply src/Google.Cloud.Functions.Hosting/HostingInternals.cs:146]: Hmm. I guess it's just about conceivable that we get given a type name that's invalid, but then a startup class replaces it with a valid one... Yup, removed.

[the remark this answers is not in the document]
Comment (jskeet) [reply src/Google.Cloud.Functions.Hosting/HostingInternals.cs:146]: (In particular, if we're *not* in that weird situation, we'll still fail clearly later.)

File: src/Google.Cloud.Functions.Hosting/EntryPoint.cs [hunk 14]
```diff
@@ -64,13 +70,21 @@ public static async Task<int> StartAsync(Assembly functionAssembly, string[] arg
...
             await host.RunAsync();
             return 0;
+
+            Type? GuessFunctionTarget()
+            {
+                var configuration = new ConfigurationBuilder()
+                    .AddFunctionsEnvironment()
+                    .AddFunctionsCommandLine(args)
+                    .Build();
+                return HostingInternals.TryGetFunctionTarget(configuration, functionAssembly);
```
[the remark this answers is not in the document]
Comment (jskeet) [reply src/Google.Cloud.Functions.Hosting/EntryPoint.cs:89]: Yup, concur.

File: src/Google.Cloud.Functions.Testing/FunctionTestServerBuilder.cs (hunk 5 above)
Comment (jskeet) [reply src/Google.Cloud.Functions.Testing/FunctionTestServerBuilder.cs:78]: Done.

File: examples/Google.Cloud.Functions.Examples.IntegrationTests/TestableDependenciesTest.cs (hunk 6 above)
Comment (jskeet) [reply examples/Google.Cloud.Functions.Examples.IntegrationTests/TestableDependenciesTest.cs:26]: Done.

File: src/Google.Cloud.Functions.Hosting/HostingInternals.cs (hunk 7 above)
Comment (jskeet) [reply src/Google.Cloud.Functions.Hosting/HostingInternals.cs:93]: Done.

File: docs/customization.md (hunk 8 above)
Comment (jskeet) [reply docs/customization.md:208]: Adding a startup is really dodgy, because it would only be used for middleware, not config or dependencies. I don't think that's something we want to support, and any previous support is accidental.

We can make this sort of check optional later on, maybe... but I think it's massively corner-case territory.

File: examples/Google.Cloud.Functions.Examples.AdvancedDependencyInjection/Function.cs (hunk 9 above)
Comment (jskeet) [reply examples/Google.Cloud.Functions.Examples.AdvancedDependencyInjection/Function.cs:78]: I've deliberately not done that because it makes it harder to read just in GitHub, which is where I'm expecting it to be read more than in VS. A VS user can always click through via the attribute itself if necessary.
